lib/index.ts

```diff
@@ -246,6 +265,91 @@ class ProgressReporter {
 		};
 	}
 
+	buildProgress(onProgress: ProgressCallback): ProgressCallback {
+		const progressRenderer = this.renderProgress;
+		let lastStep = 1;
+		let step = 0;
+		let total = 1;
+		let lastPercentage = 0;
+
+		let progressCallback: ProgressCallback = () => void 0;
+
+		return (evt: any) => {
+			let stream;
+			let id;
+			let status;
+			try {
+				({ id, stream, status } = evt);
+				if (id == null) {
+					id = '';
+				}
+				if (status == null) {
+					status = '';
+				}
+				if (stream == null) {
+					stream = '';
+				}
+
+				// Check for a `Step n/total` message on the stream
+				const streamComponents = stream.match(/^Step (\d+)\/(\d+)\s*:\s*(.+)/);
+				if (streamComponents != null) {
+					let instruction = '';
+					[, step, total, instruction] = streamComponents;
+
+					// Normalize the lastPercentage by the number of steps
+					lastPercentage = Math.floor((100.0 * (step - 1)) / total);
+
+					// The `Step n/total` is only shown at the beginning
+					// of the step. We re-create the pullProgress using the
+					// new total step value
+					if (step >= lastStep) {
+						if (instruction.toLowerCase().startsWith('from')) {
+							// If this is a from instruction, delegate progress to a new pullProgress
+							// callback
+							progressCallback = this.pullProgress(({ percentage, ...e }) => {
+								// Scale the progress from the pull step by the number of steps
+								percentage = Math.floor((step * percentage) / total);
+								percentage = lastPercentage = Math.max(
+									lastPercentage,
+									percentage,
+								);
+								return onProgress({
```

Comment:
```suggestion
								onProgress({
```
If it is all the same, removing the `return` statement reduces confusion. The definition of `ProgressCallback` indicates that nothing is returned:
```ts
export type ProgressCallback = (eventObj: any) => void;
```